Adding a new test for a empty key file.

douglasAtJoyent opened the pull request:
Adding a test for giving the client a empty key file.

This test should fail currently because of an issue that the client has.

Commit:
Adding a new test for a empty key file.

Commit:
Updaing test, add 15 more tests just trying random things with the configuration.

Commit:
Forgot to uncomment 2 of the tests, those 2 fail currently and I commented them when running the others.

tjcelaya commented on src/test/java/com/joyent/manta/archiver/ManataClientSupplierTest.java:
```diff
@@ -0,0 +1,296 @@
+                    + "JET3gPfjWkcuDOCs6zcaj/anlr/PxdSxYw4Zu21YKZGY70ceBEJgtX5PcrgmpH7R\n"
+                    + "99+DOAhlNqwK5X+4Foo3tErRveaTczgz38G17IfoKJ7rKYHQt1o+bQ==\n" + "-----END RSA PRIVATE KEY-----";
+            bw.write(validKey);
+        } catch (Exception e) {
```
These two lines just add noise and are not needed since the exception is being rethrown without any modification, please avoid this pattern in the future:

```java
} catch (Exception e) {
  throw e;
```

tjcelaya commented on src/test/java/com/joyent/manta/archiver/ManataClientSupplierTest.java:
```diff
@@ -0,0 +1,296 @@
+        }
+        System.setProperty(MANTA_KEY_PATH_PROPERTY, keyFile.getAbsolutePath());
+        MantaClient client = MANTA_CLIENT_SUPPLIER.get();
+        Stream<MantaObject> mor = client.listObjects("/" + System.getProperty(MANTA_USER) + "/stor");
```
Two things here:
- This stops being a unit test the moment it attempts to contact an external system. This kind of test belongs in `src/integration` if we intend to keep it.
- Streams tie up client connections so each call to `MantaClient#listObjects` will diminish the connections available in the client's connection pool. If you look at how `MantaTransferClient` you'll notice the supplier is only ever used to build a single client which is reused. The `MantaClient` class is thread-safe and reusing a single instance is encouraged. Creating a new client for each test case both masks the connection pool leak and skips using the client in the way that it is typically utilized.

tjcelaya commented on src/test/java/com/joyent/manta/archiver/ManataClientSupplierTest.java:
```diff
@@ -0,0 +1,296 @@
+
+    @Test(expectedExceptions = MantaClientHttpResponseException.class, expectedExceptionsMessageRegExp = ".*HTTP request failed to:.*")
+    public void invalidPassword() throws IOException {
+        System.setProperty(MANTA_USER_PROPERTY, "invalid");
```
This test is about password but this line seems to be setting the username? Changing it to "manta.password" triggers this exception which is probably what you expected to be testing:

```
com.joyent.http.signature.KeyLoadException: Loaded key is not encrypted but a password was supplied.
```

Either way, password are used when loading keys, I have no idea why you would expect a `MantaClientHttpResponseException` for a bad password. You'll also need an encrypted private key to properly test the password setting.

tjcelaya commented on src/test/java/com/joyent/manta/archiver/ManataClientSupplierTest.java:
```diff
@@ -0,0 +1,296 @@
+import com.joyent.manta.exception.ConfigurationException;
+import com.joyent.manta.exception.MantaClientHttpResponseException;
+
+import jline.internal.Log;
```
This is an immediate red flag for me, why are we using a class in a package with `internal` in the name? My editor tells me the `private static Logger` declared in the class is never used, we should go over how your editor is configured and make sure it shows warnings for unused variables.

douglasAtJoyent replied:
I had removed the logging lines, I was using it earlier and forgot to remove this.

douglasAtJoyent replied:
Actually I take that back, Line 168, Log.info(it.next());  But I will remove it. Plus I think the import is probably incorrect, that I think I just let my IDE import.

tjcelaya commented on src/test/java/com/joyent/manta/archiver/ManataClientSupplierTest.java:
```diff
@@ -0,0 +1,296 @@
+    public static final String MANTA_ENCRYPTION_AUTH_MODE = "manta.encryption_auth_mode";
+    public static final String MANTA_ENCRYPTION_KEY_PATH = "manta.encryption_key_path";
+    public static final String MANTA_ENCRYPTION_KEY_BYTES_64 = "manta.encryption_key_bytes_base";
+    private static final String base_dir = String.format("/%s", System.getenv(MANTA_USER));
```
This fails to account for subusers, I think you're looking for `com.joyent.manta.client.MantaUtils#parseAccount`

tjcelaya reviewed (changes requested):
Some of these tests could be useful but overall I'm not really sure what to make of the latest additions to this PR. I was under the impression we had discussed adding unit tests that were focused on misusing the key-related configuration options but it appears the scope has changed drastically.

I'd be happy to go over general principles for unit testing. Each test case should have a clear purpose and clearly communicate the system under test.

16 of the 18 tests in this class are failing for a variety of reasons. Should they all be passing, all be failing?

Perhaps we should pair up to define a set of test cases and work through implementing them together?

douglasAtJoyent commented on src/test/java/com/joyent/manta/archiver/ManataClientSupplierTest.java:
```diff
@@ -0,0 +1,296 @@
+
+
+    @BeforeMethod
+    private void setParameters() {
```
Yes, it assumes that you have a valid setup, I assumed that would be true since other tests also assume that you have a valid setup and a directory in the user's stor directory created.

tjcelaya replied:
Which other tests are you referencing? Only the tests in the `src/integration` folder depend on a properly-configured environment. None of the other tests in `src/test` require valid credentials. Try running `mvn clean test` in a shell without any of the `MANTA_*` variables set.

douglasAtJoyent replied:
Sorry, I think I just assumed they did since the first time I ran them they failed because a folder was not created in my /stor directory.

tjcelaya replied:
Please be more specific, are you talking about TransferManagerIT? I get this exception when I run the test so I'm assuming that's the one:

```
com.joyent.manta.archiver.TransferClientException: Remote path does not exist: /tomas.celaya/stor/manta-archiver-test/
```

Unfortunately this seems to be a problem with how that test was designed, it should actually be creating the directory during test setup in the same fashion as the integration tests in java-manta.

douglasAtJoyent replied:
How would I create a client without having the environment setup? If I try to create an instance of the client with no environmental variables set I get an exception.

douglasAtJoyent commented on src/test/java/com/joyent/manta/archiver/ManataClientSupplierTest.java:
```diff
@@ -0,0 +1,296 @@
+    private static Logger LOG = LoggerFactory.getLogger(ManataClientSupplierTest.class);
+
+    private static final MantaClientSupplier MANTA_CLIENT_SUPPLIER = new MantaClientSupplier();
+    // Seems silly, but what if they change?
```
Agreed, I will change them, I put them there more to have a easy reference.

Commit:
Updating tests according to feedback.

Commit:
1. Adding key resource files.

2. Changing the test to read files from resource.
3. Removed some unused constants.
4. Changed one of the test names. Had it as testSomething chnaged it to
basicTest.

Commit:
Removing unneeded test.

tjcelaya committed:
Minor formatting fixes

tjcelaya commented:
Sorry this took so long to get to. I pushed a commit with some small formatting fixes to make the annotation arguments more legible and switched from `KeyLoadException` to `ConfigurationException` (since the `KeyLoadException` is caught and wrapped by `MantaClient`).

Thank you for contributing!

tjcelaya commented:
(I'll leave the branch for you to delete if you want. You can reuse it in the future but it can lead to confusion if you don't merge master back into it.)
